File: engine/scanner.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sort"
	"strings"
)

// finding is the shared contract with the Python scanner: the exact JSON keys
// the dashboard consumes.
type finding struct {
	Severity     string                   `json:"severity"`
	Title        string                   `json:"title"`
	Endpoint     string                   `json:"endpoint"`
	Detail       string                   `json:"detail"`
	Reproduction string                   `json:"reproduction"`
	Evidence     []map[string]interface{} `json:"evidence"`
}

var severityOrder = map[string]int{"CRITICAL": 0, "HIGH": 1, "MEDIUM": 2, "LOW": 3}

func run(base, outJSON, outHTML string) error {
	client := &http.Client{Timeout: 10e9} // 10s
	base = strings.TrimRight(base, "/")

	spec, err := fetchSpec(client, base)
	if err != nil {
		return err
	}
	userA, err := registerUser(client, base, "alice")
	if err != nil {
		return err
	}
	userB, err := registerUser(client, base, "bob")
	if err != nil {
		return err
	}
	fmt.Printf("[*] registered alice (id=%v) and bob (id=%v)\n", userA["id"], userB["id"])

	endpoints := allEndpoints(spec)
	// deterministic order for reproducible output
	sort.Slice(endpoints, func(i, j int) bool {
		if endpoints[i].Path == endpoints[j].Path {
			return endpoints[i].Method < endpoints[j].Method
		}
		return endpoints[i].Path < endpoints[j].Path
	})
	fmt.Printf("[*] found %d operation(s)\n", len(endpoints))

	aToken, _ := userA["token"].(string)

	var findings []finding
	seenBola := map[string]bool{}
	for _, ep := range endpoints {
		findings = append(findings, checkBola(client, base, ep, userA, userB, aToken, seenBola)...)
		findings = append(findings, checkMassAssignment(client, base, ep, aToken)...)
		findings = append(findings, checkBFLA(client, base, ep, aToken)...)

		_, hasSchema := schemaPaths(ep.Op)
		if hasSchema || opHasSecurity(ep.Op) {
			// baseline = a 200 response for A's own resource (or direct call)
			filled, body := resolveBaseline(client, base, ep, userA, aToken)
			if hasSchema && filled != "" && body != nil {
				findings = append(findings, checkExposure(ep, body, base, filled, aToken)...)
			}
			if opHasSecurity(ep.Op) && filled != "" {
				findings = append(findings, checkMissingAuth(client, ep, base, filled)...)
			}
		}
	}

	sort.SliceStable(findings, func(i, j int) bool {
		return severityOrder[findings[i].Severity] < severityOrder[findings[j].Severity]
	})

	if err := writeFindings(outJSON, findings); err != nil {
		return err
	}
	if err := writeReport(outHTML, spec, findings, base); err != nil {
		return err
	}

	fmt.Printf("[*] %d finding(s):\n", len(findings))
	for _, f := range findings {
		fmt.Printf("    %-9s %s  ->  %s\n", f.Severity, f.Title, f.Endpoint)
	}
	fmt.Printf("[*] wrote %s\n", outJSON)
	fmt.Printf("[*] wrote %s\n", outHTML)
	return nil
}

func registerUser(client *http.Client, base, username string) (map[string]interface{}, error) {
	payload, _ := json.Marshal(map[string]string{"username": username})
	resp, err := client.Post(base+"/register", "application/json", bytes.NewReader(payload))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("register returned %d", resp.StatusCode)
	}
	var v map[string]interface{}
	dec := json.NewDecoder(resp.Body)
	dec.UseNumber()
	if err := dec.Decode(&v); err != nil {
		return nil, err
	}
	return v, nil
}

// doRequest performs a request of any method. Write methods get a minimal JSON
// body. An empty token means no Authorization header is sent.
func doRequest(client *http.Client, method, url, token string) (int, interface{}, error) {
	var body io.Reader
	switch method {
	case http.MethodPost, http.MethodPut, http.MethodPatch:
		body = bytes.NewReader([]byte("{}"))
	}
	req, err := http.NewRequest(method, url, body)
	if err != nil {
		return 0, nil, err
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}
	resp, err := client.Do(req)
	if err != nil {
		return 0, nil, err
	}
	defer resp.Body.Close()
	data, _ := io.ReadAll(resp.Body)
	var parsed interface{}
	if err := json.Unmarshal(data, &parsed); err != nil {
		parsed = nil
	}
	return resp.StatusCode, parsed, nil
}

// resolveBaseline finds a 200 response for A's own resource (substituting A's
// identifiers into a single path param) or, for paths with no params, a direct
// call. Returns the filled path and the response body, or ""/nil if none.
func resolveBaseline(client *http.Client, base string, ep endpoint, a map[string]interface{}, aToken string) (string, interface{}) {
	params := pathParams(ep.Path)
	if len(params) == 1 {
		for _, m := range markersFromUser(a) {
			cand := strings.Replace(ep.Path, "{"+params[0]+"}", m.Str, 1)
			status, body, err := doRequest(client, ep.Method, base+cand, aToken)
			if err == nil && status == http.StatusOK {
				return cand, body
			}
		}
		return "", nil
	}
	// no path params: only read endpoints get a baseline (avoid mutating writes)
	if ep.Method != http.MethodGet && ep.Method != http.MethodHead {
		return "", nil
	}
	status, body, err := doRequest(client, ep.Method, base+ep.Path, aToken)
	if err == nil && status == http.StatusOK {
		return ep.Path, body
	}
	return "", nil
}

// checkBola is the cross-principal differential test (any HTTP method): substitute
// B's identifier into the path while presenting A's token, and flag if B-owned
// data comes back.
func checkBola(client *http.Client, base string, ep endpoint, a, b map[string]interface{}, aToken string, seen map[string]bool) []finding {
	params := pathParams(ep.Path)
	if len(params) != 1 {
		return nil
	}
	key := ep.Method + " " + ep.Path
	bMarkers := markersFromUser(b)
	var out []finding
	for _, sub := range bMarkers {
		filled := strings.Replace(ep.Path, "{"+params[0]+"}", sub.Str, 1)
		status, body, err := doRequest(client, ep.Method, base+filled, aToken)
		if err != nil || status != http.StatusOK {
			continue
		}
		for _, m := range bMarkers {
			if containsValue(body, m.Str) {
				if seen[key] {
					break
				}
				seen[key] = true
				out = append(out, finding{
					Severity: "CRITICAL",
					Title:    "Broken Object-Level Authorization (BOLA / IDOR)",
					Endpoint: key,
					Detail: fmt.Sprintf(
						"Account '%s' (token A) called %s and received data owned by '%s' (field %s=%s). No ownership check is enforced.",
						str(a["username"]), filled, str(b["username"]), m.Key, repr(m.Raw)),
					Reproduction: curlRepro(ep.Method, base, filled, aToken),
					Evidence:     []map[string]interface{}{{m.Key: m.Raw}},
				})
				break
			}
		}
	}
	return out
}

// checkExposure recursively diffs the actual response paths against the declared
// schema paths, flagging undeclared fields (including nested ones).
func checkExposure(ep endpoint, body interface{}, base, filled, aToken string) []finding {
	declared, ok := schemaPaths(ep.Op)
	if !ok {
		return nil
	}

	var respPaths []string
	collectResponsePaths(body, "", &respPaths)
	actual := make(map[string]bool, len(respPaths))
	for _, p := range respPaths {
		actual[p] = true
	}

	var extras []string
	for p := range actual {
		if !declared[p] {
			extras = append(extras, p)
		}
	}
	sort.Strings(extras)

	var out []finding
	for _, p := range extras {
		sev := "MEDIUM"
		if sensitive(lastSegment(p)) {
			sev = "HIGH"
		}
		out = append(out, finding{
			Severity: sev,
			Title:    "Excessive Data Exposure",
			Endpoint: ep.Method + " " + ep.Path,
			Detail: fmt.Sprintf(
				"Field '%s' is returned in the response but is NOT declared in the OpenAPI response schema. The client receives more data than the contract specifies.",
				p),
			Reproduction: curlRepro(ep.Method, base, filled, aToken),
			Evidence: []map[string]interface{}{{
				"undeclared_field": p,
				"declared_fields":  sortedKeys(declared),
			}},
		})
	}
	return out
}

// checkMissingAuth flags endpoints that DECLARE a security requirement in the
// spec but return data (HTTP 200) with no Authorization header at all.
func checkMissingAuth(client *http.Client, ep endpoint, base, filled string) []finding {
	status, body, err := doRequest(client, ep.Method, base+filled, "")
	if err != nil || status != http.StatusOK {
		return nil
	}
	if body == nil {
		return nil
	}
	return []finding{{
		Severity: "HIGH",
		Title:    "Broken Authentication",
		Endpoint: ep.Method + " " + ep.Path,
		Detail: fmt.Sprintf(
			"The spec declares an API-key security requirement for %s, but it returns data with no Authorization header (HTTP 200).",
			ep.Method+" "+ep.Path),
		Reproduction: curlNoAuth(ep.Method, base, filled),
		Evidence:     []map[string]interface{}{{"http_status": 200}},
	}}
}

// checkMassAssignment injects privileged fields into write endpoints and flags
// any that accept and reflect them.
func checkMassAssignment(client *http.Client, base string, ep endpoint, aToken string) []finding {
	switch ep.Method {
	case http.MethodPost, http.MethodPut, http.MethodPatch:
	default:
		return nil
	}
	const injected = `{"username":"mallory","role":"admin","is_admin":true,"balance":999999}`
	req, err := http.NewRequest(ep.Method, base+ep.Path, bytes.NewReader([]byte(injected)))
	if err != nil {
		return nil
	}
	req.Header.Set("Content-Type", "application/json")
	if aToken != "" {
		req.Header.Set("Authorization", "Bearer "+aToken)
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		return nil
	}
	data, _ := io.ReadAll(resp.Body)
	var parsed interface{}
	if err := json.Unmarshal(data, &parsed); err != nil {
		return nil
	}
	var echoed []string
	for _, f := range []string{"role", "is_admin"} {
		if hasKey(parsed, f) {
			echoed = append(echoed, f)
		}
	}
	if len(echoed) == 0 {
		return nil
	}
	return []finding{{
		Severity: "HIGH",
		Title:    "Mass Assignment",
		Endpoint: ep.Method + " " + ep.Path,
		Detail: fmt.Sprintf(
			"The write endpoint accepts and persists privileged fields not part of its contract (%s). A client can escalate its own privileges.",
			strings.Join(echoed, ", ")),
		Reproduction: curlBody(ep.Method, base, ep.Path, aToken, injected),
		Evidence:     []map[string]interface{}{{"injected_fields": echoed}},
	}}
}

// checkBFLA flags admin-prefixed endpoints that a regular (non-admin) token can
// access — broken function-level authorization.
func checkBFLA(client *http.Client, base string, ep endpoint, aToken string) []finding {
	if !strings.HasPrefix(ep.Path, "/admin/") {
		return nil
	}
	status, _, err := doRequest(client, ep.Method, base+ep.Path, aToken)
	if err != nil || status != http.StatusOK {
		return nil
	}
	return []finding{{
		Severity: "HIGH",
		Title:    "Broken Function Level Authorization",
		Endpoint: ep.Method + " " + ep.Path,
		Detail: fmt.Sprintf(
			"A regular (non-admin) token can access %s (HTTP 200); the endpoint should be restricted to admin roles.",
			ep.Method+" "+ep.Path),
		Reproduction: curlRepro(ep.Method, base, ep.Path, aToken),
		Evidence:     []map[string]interface{}{{"role": "user", "http_status": 200}},
	}}
}

func curlRepro(method, base, path, token string) string {
	return fmt.Sprintf("curl -s -H 'Authorization: Bearer %s' '%s%s'", token, base, path)
}

func curlNoAuth(method, base, path string) string {
	return fmt.Sprintf("curl -s -X %s '%s%s'", method, base, path)
}

func curlBody(method, base, path, token, body string) string {
	return fmt.Sprintf("curl -s -X %s -H 'Authorization: Bearer %s' -H 'Content-Type: application/json' -d '%s' '%s%s'",
		method, token, body, base, path)
}

func str(v interface{}) string {
	s, _ := v.(string)
	return s
}

func sortedKeys(m map[string]bool) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

File: scanner/scanner.py
"""
SentinelAPI — the scanner engine.

Takes a target API's OpenAPI spec + base URL, registers two principals, and runs
three vulnerability classes:

  1. BOLA / IDOR            — cross-account differential test across ALL HTTP
                              methods; if B-owned data comes back under A's
                              token, the endpoint is flagged CRITICAL.
  2. Excessive data exposure — recursively diff actual response paths against
                              the schema (including nested objects).
  3. Broken authentication  — endpoints that declare a security requirement
                              but return data with no token are flagged HIGH.

Outputs:
  findings.json   — machine-readable results
  report.html     — self-contained, severity-ranked dashboard (no server needed)

Usage:
  python scanner/scanner.py --base http://127.0.0.1:8000 \
        --out-json scanner/findings.json --out-html scanner/report.html
"""
from __future__ import annotations

import argparse
import html
import json
import re
import sys
from dataclasses import dataclass, field

import httpx

SEVERITY_ORDER = {"CRITICAL": 0, "HIGH": 1, "MEDIUM": 2, "LOW": 3}
# keys whose presence in a response is plausibly a leak
SENSITIVE_HINTS = (
    "password", "passwd", "secret", "token", "ssn", "card", "cvv",
    "pan", "key", "otp", "aadhaar", "pan_number",
)

# marker fields returned by /register that identify a principal (used as the
# oracle for "is this B's data?")
IDENTIFIER_FIELDS = ("id", "username", "email", "order_id", "post_id")


@dataclass
class Finding:
    severity: str
    title: str
    endpoint: str
    detail: str
    reproduction: str
    evidence: list = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "severity": self.severity,
            "title": self.title,
            "endpoint": self.endpoint,
            "detail": self.detail,
            "reproduction": self.reproduction,
            "evidence": self.evidence,
        }


# ---------------------------------------------------------------------------
# helpers
# ---------------------------------------------------------------------------
def fetch_spec(client: httpx.Client, base: str) -> dict:
    r = client.get(f"{base}/openapi.json")
    r.raise_for_status()
    return r.json()


def register(client: httpx.Client, base: str, username: str) -> dict:
    r = client.post(f"{base}/register", json={"username": username})
    r.raise_for_status()
    return r.json()


HTTP_METHODS = {"get", "post", "put", "patch", "delete", "head", "options", "trace"}


def all_endpoints(spec: dict) -> list[tuple[str, str, dict]]:
    """Return [(method, path, operation)] for every operation in the spec."""
    out = []
    for path, methods in spec.get("paths", {}).items():
        for method, op in methods.items():
            if isinstance(op, dict) and method.lower() in HTTP_METHODS:
                out.append((method.upper(), path, op))
    return out


def path_params(path: str) -> list[str]:
    return re.findall(r"\{([^}]+)\}", path)


def contains_value(obj, target) -> bool:
    """Recursive search: is `target` present anywhere in the JSON object?"""
    if isinstance(obj, dict):
        for v in obj.values():
            if v == target or contains_value(v, target):
                return True
    elif isinstance(obj, list):
        for v in obj:
            if v == target or contains_value(v, target):
                return True
    return False


def op_has_security(op: dict) -> bool:
    return bool(op and op.get("security"))


def schema_paths(op: dict) -> set | None:
    """Return dotted paths declared in the 200 response schema, or None."""
    try:
        props = op["responses"]["200"]["content"]["application/json"]["schema"]["properties"]
    except (KeyError, TypeError):
        return None

    out: set[str] = set()

    def walk(p, prefix=""):
        for name, raw in p.items():
            path = f"{prefix}.{name}" if prefix else name
            if isinstance(raw, dict):
                if "properties" in raw:
                    walk(raw["properties"], path)
                elif isinstance(raw.get("items"), dict) and "properties" in raw["items"]:
                    walk(raw["items"]["properties"], path)
                else:
                    out.add(path)
            else:
                out.add(path)

    walk(props)
    return out


def collect_response_paths(obj, prefix="", out=None):
    """Return dotted paths to every leaf in a JSON response."""
    if out is None:
        out = []
    if isinstance(obj, dict):
        for k, v in obj.items():
            p = f"{prefix}.{k}" if prefix else k
            if isinstance(v, (dict, list)):
                collect_response_paths(v, p, out)
            else:
                out.append(p)
    elif isinstance(obj, list):
        for v in obj:
            collect_response_paths(v, prefix, out)
    return out


def last_segment(p: str) -> str:
    return p.rsplit(".", 1)[-1]


def has_key(obj, key: str) -> bool:
    if isinstance(obj, dict):
        if key in obj:
            return True
        return any(has_key(v, key) for v in obj.values())
    if isinstance(obj, list):
        return any(has_key(v, key) for v in obj)
    return False


def sensitive(key: str) -> bool:
    k = key.lower()
    return any(h in k for h in SENSITIVE_HINTS)


def curl_repro(method: str, base: str, path: str, token: str) -> str:
    return f"curl -s -H 'Authorization: Bearer {token}' '{base}{path}'"


def curl_no_auth(method: str, base: str, path: str) -> str:
    return f"curl -s -X {method} '{base}{path}'"


def curl_body(method: str, base: str, path: str, token: str, body: str) -> str:
    return (f"curl -s -X {method} -H 'Authorization: Bearer {token}' "
            f"-H 'Content-Type: application/json' -d '{body}' '{base}{path}'")


def do_request(client, method, url, token):
    """Request of any method. Write methods get a minimal JSON body; an empty
    token means no Authorization header is sent."""
    headers = {}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    kwargs = {"headers": headers}
    if method in ("POST", "PUT", "PATCH"):
        kwargs["json"] = {}
    r = client.request(method, url, **kwargs)
    try:
        return r.status_code, r.json()
    except ValueError:
        return r.status_code, None
def check_bola(client, base, endpoint, user_a, user_b, seen) -> list[Finding]:
    """Cross-account differential test for broken object-level authorization (any method)."""
    findings = []
    method, path, _op = endpoint
    params = path_params(path)
    if len(params) != 1:
        return findings

    b_markers = {k: user_b[k] for k in IDENTIFIER_FIELDS if k in user_b}
    for marker_key, marker_val in b_markers.items():
        filled = path.replace("{" + params[0] + "}", str(marker_val))
        try:
            status, body = do_request(client, method, base + filled, user_a["token"])
        except httpx.HTTPError as exc:
            print(f"  [!] request error on {filled}: {exc}", file=sys.stderr)
            continue
        if status != 200:
            continue

        for mk, mv in b_markers.items():
            if mv is not None and contains_value(body, mv):
                key = ("BOLA", method, path)
                if key in seen:
                    continue
                seen.add(key)
                findings.append(Finding(
                    severity="CRITICAL",
                    title="Broken Object-Level Authorization (BOLA / IDOR)",
                    endpoint=f"{method} {path}",
                    detail=(
                        f"Account '{user_a['username']}' (token A) called {filled} "
                        f"and received data owned by '{user_b['username']}' "
                        f"(field {mk}={mv!r}). No ownership check is enforced."
                    ),
                    reproduction=curl_repro(method, base, filled, user_a["token"]),
                    evidence=[{mk: mv}],
                ))
                break
    return findings


def check_exposure(endpoint, body, base, filled, user_a) -> list[Finding]:
    """Recursively diff actual response paths vs declared schema paths."""
    method, path, op = endpoint
    declared = schema_paths(op)
    if declared is None:
        return []

    actual = set(collect_response_paths(body))
    extras = actual - declared
    findings = []
    for key in sorted(extras):
        sev = "HIGH" if sensitive(last_segment(key)) else "MEDIUM"
        findings.append(Finding(
            severity=sev,
            title="Excessive Data Exposure",
            endpoint=f"{method} {path}",
            detail=(
                f"Field '{key}' is returned in the response but is NOT declared "
                "in the OpenAPI response schema. The client receives more data "
                "than the contract specifies."
            ),
            reproduction=curl_repro(method, base, filled, user_a["token"]),
            evidence=[{"undeclared_field": key, "declared_fields": sorted(declared)}],
        ))
    return findings


def check_missing_auth(client, endpoint, base, filled) -> list[Finding]:
    """Flag endpoints that declare security but return data with no token."""
    method, path, _op = endpoint
    try:
        status, body = do_request(client, method, base + filled, "")
    except httpx.HTTPError:
        return []
    if status != 200 or body is None:
        return []
    return [Finding(
        severity="HIGH",
        title="Broken Authentication",
        endpoint=f"{method} {path}",
        detail=(
            f"The spec declares an API-key security requirement for {method} {path}, "
            "but it returns data with no Authorization header (HTTP 200)."
        ),
        reproduction=curl_no_auth(method, base, filled),
        evidence=[{"http_status": 200}],
    )]


def check_mass_assignment(client, base, endpoint, a_token) -> list[Finding]:
    """Inject privileged fields into write endpoints; flag if accepted & echoed."""
    method, path, _op = endpoint
    if method not in ("POST", "PUT", "PATCH"):
        return []
    injected = '{"username":"mallory","role":"admin","is_admin":true,"balance":999999}'
    try:
        r = client.request(method, base + path, content=injected, headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {a_token}",
        })
    except httpx.HTTPError:
        return []
    if r.status_code not in (200, 201):
        return []
    try:
        body = r.json()
    except ValueError:
        return []
    echoed = [f for f in ("role", "is_admin") if has_key(body, f)]
    if not echoed:
        return []
    return [Finding(
        severity="HIGH",
        title="Mass Assignment",
        endpoint=f"{method} {path}",
        detail=(
            "The write endpoint accepts and persists privileged fields not part "
            f"of its contract ({', '.join(echoed)}). A client can escalate its "
            "own privileges."
        ),
        reproduction=curl_body(method, base, path, a_token, injected),
        evidence=[{"injected_fields": echoed}],
    )]


def check_bfa(client, base, endpoint, a_token) -> list[Finding]:
    """Flag admin-prefixed endpoints a regular (non-admin) token can access."""
    method, path, _op = endpoint
    if not path.startswith("/admin/"):
        return []
    try:
        status, _ = do_request(client, method, base + path, a_token)
    except httpx.HTTPError:
        return []
    if status != 200:
        return []
    return [Finding(
        severity="HIGH",
        title="Broken Function Level Authorization",
        endpoint=f"{method} {path}",
        detail=(
            f"A regular (non-admin) token can access {method} {path} (HTTP 200); "
            "the endpoint should be restricted to admin roles."
        ),
        reproduction=curl_repro(method, base, path, a_token),
        evidence=[{"role": "user", "http_status": 200}],
    )]


# ---------------------------------------------------------------------------
# report rendering (self-contained HTML — no server required)
# ---------------------------------------------------------------------------
def render_html(spec: dict, findings: list[Finding], base: str) -> str:
    title = spec.get("info", {}).get("title", "Target API")
    cards = []
    for f in findings:
        badge = {
            "CRITICAL": "#e5484d",
            "HIGH": "#f76b15",
            "MEDIUM": "#ffb224",
            "LOW": "#3e63dd",
        }.get(f.severity, "#8d8d8d")
        ev = html.escape(json.dumps(f.evidence, indent=2))
        cards.append(f"""
        <div class="card">
          <div class="card-head">
            <span class="badge" style="background:{badge}">{html.escape(f.severity)}</span>
            <span class="title">{html.escape(f.title)}</span>
          </div>
          <div class="endpoint">{html.escape(f.endpoint)}</div>
          <p class="detail">{html.escape(f.detail)}</p>
          <pre class="curl">{html.escape(f.reproduction)}</pre>
          <details><summary>evidence</summary><pre>{ev}</pre></details>
        </div>""")

    counts = {}
    for f in findings:
        counts[f.severity] = counts.get(f.severity, 0) + 1
    summary = " · ".join(f"{k}: {v}" for k, v in
                         sorted(counts.items(), key=lambda kv: SEVERITY_ORDER.get(kv[0], 9)))

    return f"""<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>SentinelAPI — Scan Report</title>
<style>
  :root {{ color-scheme: dark; }}
  body {{ margin:0; font-family:-apple-system,Segoe UI,Roboto,Helvetica,Arial,sans-serif;
         background:#0d1117; color:#e6edf3; }}
  header {{ padding:28px 36px; border-bottom:1px solid #21262d;
           background:linear-gradient(180deg,#161b22,#0d1117); }}
  h1 {{ margin:0 0 6px; font-size:22px; }}
  .sub {{ color:#8b949e; font-size:13px; }}
  .stats {{ margin-top:12px; font-size:14px; color:#79c0ff; }}
  main {{ padding:24px 36px; display:grid; gap:16px; }}
  .card {{ background:#161b22; border:1px solid #30363d; border-radius:10px; padding:16px; }}
  .card-head {{ display:flex; align-items:center; gap:10px; }}
  .badge {{ font-size:11px; font-weight:700; letter-spacing:.04em; color:#fff;
           padding:3px 9px; border-radius:999px; }}
  .title {{ font-weight:600; font-size:15px; }}
  .endpoint {{ color:#79c0ff; font-family:ui-monospace,SFMono-Regular,Menlo,monospace;
              font-size:13px; margin:8px 0; }}
  .detail {{ color:#c9d1d9; font-size:13px; line-height:1.5; margin:8px 0; }}
  .curl {{ background:#0d1117; border:1px solid #30363d; border-radius:6px;
          padding:10px; font-size:12px; color:#7ee787; overflow-x:auto; }}
  details {{ color:#8b949e; font-size:12px; margin-top:8px; }}
  details pre {{ background:#0d1117; padding:8px; border-radius:6px; overflow-x:auto; }}
  .empty {{ color:#8b949e; padding:40px; text-align:center; }}
</style>
</head>
<body>
<header>
  <h1>🛡️ SentinelAPI — Scan Report</h1>
  <div class="sub">Target: {html.escape(title)} · {html.escape(base)}</div>
  <div class="stats">{html.escape(summary) or 'no findings'}</div>
</header>
<main>
  {''.join(cards) or '<div class="empty">No vulnerabilities detected ✓</div>'}
</main>
</body>
</html>"""


# ---------------------------------------------------------------------------
# main
# ---------------------------------------------------------------------------
def main() -> int:
    ap = argparse.ArgumentParser(description="SentinelAPI scanner")
    ap.add_argument("--base", required=True, help="target API base URL")
    ap.add_argument("--out-json", default="findings.json")
    ap.add_argument("--out-html", default="report.html")
    args = ap.parse_args()

    base = args.base.rstrip("/")
    client = httpx.Client(timeout=10, follow_redirects=False)

    spec = fetch_spec(client, base)
    user_a = register(client, base, "alice")
    user_b = register(client, base, "bob")
    print(f"[*] registered alice (id={user_a['id']}) and bob (id={user_b['id']})")

    endpoints = all_endpoints(spec)
    print(f"[*] found {len(endpoints)} operation(s)")

    a_token = user_a["token"]

    def resolve_baseline(ep):
        method, path, _op = ep
        params = path_params(path)
        if len(params) == 1:
            for k in IDENTIFIER_FIELDS:
                if k not in user_a:
                    continue
                cand = path.replace("{" + params[0] + "}", str(user_a[k]))
                try:
                    status, body = do_request(client, method, base + cand, a_token)
                except httpx.HTTPError:
                    continue
                if status == 200:
                    return cand, body
            return "", None
        if method not in ("GET", "HEAD"):
            return "", None  # avoid mutating writes
        try:
            status, body = do_request(client, method, base + path, a_token)
        except httpx.HTTPError:
            return "", None
        return (path, body) if status == 200 else ("", None)

    findings: list[Finding] = []
    seen: set = set()
    for ep in endpoints:
        findings += check_bola(client, base, ep, user_a, user_b, seen)
        findings += check_mass_assignment(client, base, ep, a_token)
        findings += check_bfa(client, base, ep, a_token)

        declared = schema_paths(ep[2])
        if declared is not None or op_has_security(ep[2]):
            filled, body = resolve_baseline(ep)
            if declared is not None and filled and body is not None:
                findings += check_exposure(ep, body, base, filled, user_a)
            if op_has_security(ep[2]) and filled:
                findings += check_missing_auth(client, ep, base, filled)

    findings.sort(key=lambda f: SEVERITY_ORDER.get(f.severity, 9))

    with open(args.out_json, "w") as fh:
        json.dump([f.to_dict() for f in findings], fh, indent=2)
    with open(args.out_html, "w") as fh:
        fh.write(render_html(spec, findings, base))

    print(f"[*] {len(findings)} finding(s):")
    for f in findings:
        print(f"    {f.severity:<9} {f.title}  ->  {f.endpoint}")
    print(f"[*] wrote {args.out_json}")
    print(f"[*] wrote {args.out_html}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
